src/wpool/mongoose_gun_worker.erl

```diff
@@ -0,0 +1,262 @@
+                        headers :: gun:headers(),
+                        acc = <<>> :: binary()}).
+
+-type response_data() :: #response_data{}.
+
+-type request() :: {request,
+                    Path :: iodata(),
+                    Method :: iodata(),
+                    Headers :: gun:headers(),
+                    Query :: iodata(),
+                    Retries :: non_neg_integer(),
+                    Timeout :: non_neg_integer()}.
+
+-include("mongoose_logger.hrl").
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link(Destination, Options) ->
+    gen_server:start_link(?MODULE, {Destination, Options}, []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec init({{inet:hostname() | inet:ip_address(), inet:port_number()}, gun:opts()}) ->
+    {ok, State :: #state{}, {continue, init}}.
+init({{Host, Port}, Opts}) ->
+    {H, _P, O} = parse_uri(Host),
+    {ok, #state{host = H, port = Port, opts = maps:merge(O, Opts), requests = #{}},
+     {continue, init}};
+init({Host, Opts}) ->
+    {H, P, O} = parse_uri(Host),
+    {ok, #state{host = H, port = P, opts = maps:merge(O, Opts), requests = #{}},
+     {continue, init}}.
+
+handle_continue(init, State) ->
+    NewState = open_connection(State),
+    {noreply, NewState}.
+
+-spec handle_call(Request :: request(), From :: {pid(), Tag :: term()}, State :: #state{}) ->
+    {noreply, NewState :: #state{}}.
+handle_call({request, FullPath, Method, Headers, Query, _Retries, Timeout} = Request,
+            From,
+            #state{requests = Requests} = State) ->
+
+    Now = erlang:monotonic_time(millisecond),
+    LHeaders = lowercase_headers(Headers),
+
+    {StreamRef, TRef} = queue_request(FullPath, Method, LHeaders, Query, Timeout, State#state.pid),
+    NewRequests = Requests#{StreamRef => {Request, #response_data{from = From,
+                                                                  timestamp = Now,
+                                                                  timeout_timer = TRef}}},
+    {noreply, State#state{requests = NewRequests}}.
+
+-spec handle_cast(any(), State :: #state{}) -> {noreply, State :: #state{}}.
+handle_cast(_R, State) ->
+    {noreply, State}.
+
+%% @doc Handles gun messages.
+%% `gun_response' is a message that informs about the response to the request
+%% and provides the response headers and status. If there is no body, the `fin'
+%% flag is set.
+handle_info({gun_response, ConnPid, StreamRef, fin, Status, Headers},
+            #state{pid = ConnPid, requests = Requests} = State) ->
+    case maps:get(StreamRef, Requests, undefined) of
+        undefined ->
+            {noreply, State};
+        {_Req, ResData} ->
+            timer:cancel(ResData#response_data.timeout_timer),
```

Comment:
Please _do not use_ the `timer` module at all, it's a global gen_server routing all timer requests, that is, a terrible bottleneck. Use bifs as defined in the `erlang` module.